### ISSUE TITLE
Video capture feature in Allegro 5

### ISSUE THREAD

**Issue by KaleviKolttonen** [liballeg/allegro5#903]
Would it be possible include a video capture feature in Allegro 5 library?

**Comment by beoran** [comment]
With the current functionality, you can already capture audio, and you
could save frames from your game to memory using the FS interface and
al_save_bitmap. You would have to still implement compression yourself. I
admit easy video capture would be a nice feature, but you might need to
implement it yourself if you need it.

**Comment by dos1** [comment]
I would have guessed that this ticket is about capturing the video for instance from the webcam :)

**Comment by KaleviKolttonen** [comment]
No, it is not about webcams. There exists a BBC Micro (an old 8-bit computer) emulator called B-Em. I was wondering how one could implement video capturing in that emulator. It turned out that the newest versions use Allegro 5, so one of the B-Em developers suggested that the video capture feature would be most useful inside the Allegro 5 library. That way potentially all Allegro 5 applications could benefit from it, having a pretty simple API to do the capturing.

**Comment by SteveFosdick** [comment]
beoran, first let me introduce myself as the B-Em developer KaleviKolttonen was in discussion with.

The reason I suggested this might be a good additional to the Allegro library is because the use case is to record game play in order to demonstrate to others. (Although B-Em is emulating a BBC Micro that was, itself a game platform in the 1980s as well as other things).  It seemed likely to me that others who play games which use the Allegro library may also find this useful.

When you say it is already possible to capture audio which part of the Allegro 5 did you have in mind.  I have checked out the audio recorder feature (http://liballeg.org/a5docs/trunk/audio.html#audio-recording) but that specifically says it captures the default system recording input which is unlikely to be a copy of what is being heard.  For B-Em, we are trying to be cross-platform, hence using Allegro rather than using platform specific APIs, and thus altering the sound routing on the OS isn't something we want to be doing.  Were you thinking of al_set_mixer_postprocess_callback?

**Comment by beoran** [comment]
Yes, I think that by playing around with the mixer API it should be
possible to capture the sound as it gets played by storing the buffers that
Allegro passes to your callbacks. Of course I admit it is easier said than
done.
